doc/gpa.1
.TH "gpa" "1" "" "" "User Commands"
.nh
.SH "NAME"
gpa \- graphical frontend for the GNU Privacy Guard

.SH "SYNOPSIS"
.B gpa
.RB [ "OPTION(S)" ]
.RI [ "FILE(S)" ]

.SH "DESCRIPTION"
The GNU Privacy Assistant (GPA) is a graphical user interface for the GNU
Privacy Guard (GnuPG).
.PP
\fBgpa\fP can be used to encrypt, decrypt, and sign files, to verify
signatures and to manage the private and public keys.

.SH "OPTIONS"
.TP
.B \-c, \-\-clipboard
Open the clipboard.  This is the \fIdefault\fP action if no arguments are
given and a key has already been created.
.TP
.B \-C, \-\-card
Start with the card-manager open.
.TP
.B \-d, \-\-daemon
Only start the UI server and no user interface.
.TP
.B \-\-no-remote
Do not connect to a running instance but start a new one.  This can
also be used to not start an UI server.
.TP
.B \-\-debug-edit-fsm
Debug the Finite State Machine (FSM).
.TP
.B \-\-disable\-ticker
Disable ticker used for card operations.
.TP
.B \-\-disable-x509
Disable support for X.509.
.TP
.B \-\-enable-logging
Enable logging on Win32 systems.
.TP
.B \-f, \-\-files
Start with the file-manager open. This is the \fIdefault\fP if one or more
\fIFILE(S)\fP are added to the command arguments.
.TP
.B \-k, \-\-keyring
Start with the keyring editor. This is the \fIdefault\fP for a new
installation.
.TP
.B \-o, \-\-options=\fIFILE\fP
Read options from the specified file instead of \fI~/.gnupg/gpa.conf\fP.
.TP
.B \-s, \-\-settings
Open the settings dialog.
.TP
.B \-v, \-\-version
Print version information and exit.
.TP
.B \-\-help\-gtk
Print options related to GTK. See also
.BR gtk-options "(7)."
.TP
.B \-?, \-\-help, \-\-help\-all
Print usage information and exit.

.SH FILES
.TP
.I ~/.gnupg/gpa.conf
Standard user configuration file read by \fBgpa\fP on startup.
.TP
.I ~/.gnupg/S.uiserver
Socket listening for commands to the UI server.

.SH "BUGS"
Please report bugs to
\fI\%http://bugs.gnupg.org/\fR.

.SH AUTHORS
.B gpa
was mainly written by \fBBernhard Herzog\fP, \fBMarcus Brinkmann\fP,
\fBMiguel Coca\fP, and \fBWerner Koch\fP.  See the \fIhelp menu\fR and
\fI\%http://gitstats.gnupg.org/gpa/authors.html\fR for a detailed
list.
.PP
This manual page was written by \fBArthur de Jong\fP <\&adejong@debian.org\&>
and \fBDaniel Leidert\fP <\&daniel.leidert@wgdd.de\&> for the Debian GNU/Linux
system (but may be used by others).

.SH "SEE ALSO"
.BR gpg (1),
.BR gpg2 (1),
.BR gpgconf (1),
.BR gpgsm (1),
.BR gtk-options (7)


NEWS
Noteworthy changes in version 0.9.8 (unreleased)
------------------------------------------------

Noteworthy changes in version 0.9.7 (2014-12-12)
------------------------------------------------

 * Support sending keys for GnuPG 2.1.


Noteworthy changes in version 0.9.6 (2014-11-21)
------------------------------------------------

 * Support keyserver operations for GnuPG 2.1.

 * Implement the IMPORT_FILES server command.

 * New "Refresh Key" action in the key manager's context menu.


Noteworthy changes in version 0.9.5 (2014-09-01)
------------------------------------------------

 * GPA now starts with the UI server enabled and tests on startup
   whether such a server is already running to open that one instead
   of launching a second instance.

 * GPA is now aware of ECC keys.

 * Improved detection of CMS objects (which are used by S/MIME) and
   detached OpenPGP signatures.

 * Allow import and export of X.509 certificates.  Allow backup of
   X.509 keys.

 * The key creation date is now displayed in the key listing.

 * Armored detached signature files are now created with an ".asc"
   suffix and not with ".sig".

 * The GnuPG home directory is now detected using the gpgconf tool.

 * Added launch-gpa wrapper for Windows.

 * Fixed several bugs leading to crashs.


Noteworthy changes in version 0.9.4 (2013-05-01)
------------------------------------------------

 * Added scrollbars to the verification result window.

 * Improved searching in the key listing.

 * Now uses the native theme under Windows.


Noteworthy changes in version 0.9.3 (2012-08-08)
------------------------------------------------

 * Allow searching in the keylist.

 * Collected bug fixes.


Noteworthy changes in version 0.9.2 (2012-05-02)
------------------------------------------------

 * Adjust server mode to modern Libassuan.

 * Add options --enable-logging for W32.

 * Add options --gpg-binary, --gpgsm-binary and --debug-edit-fsm.

 * Properly process CMS data in the clipboard and with the server's
   VERIFY_FILES and DECRYPT_FILES commands.

 * Minor code cleanups.


Noteworthy changes in version 0.9.1 (2012-04-18)
------------------------------------------------

 * The key selection dialogs for encryption and signing do not anymore
   list expired, revoked or otherwise invalid keys.

 * If no recipients are given to the server, a generic key selection
   dialog is now used.

 * Now works with Libassuan 2.x.

 * The card manager now displays the ATR for an unknown card.


Noteworthy changes in version 0.9.0 (2009-06-20)
------------------------------------------------

 * Added a smartcard manager.

 * GPA now requires GnuPG-2.

 * X.509 support is now always enabled.

 * Major internal cleanups.  More to follow soon.


Noteworthy changes in version 0.8.0 (2008-09-04)
------------------------------------------------

 * Add basic UI server mode and option --daemon.

 * GPA now supports direct crypto operations to and from the
   clipboard, and features a simple text edit area as well.

 * GPA supports manipulating the backend configuration through
   gpg-conf.

 * GPA has now basic support for X.509; use the command line switch
   --cms to enable this.

 * The default keyserver is now taken from gpg.conf and not from
   gpa.conf.


Noteworthy changes in version 0.7.6 (2007-05-24)
------------------------------------------------

 * Czech translation by Zdenek Hatas.

 * Russian translation by Maxim Britov.

 * Files may now be dropped onto the file manager window.


Noteworthy changes in version 0.7.5 (2007-02-26)
------------------------------------------------

 * Allow setting a password if it was empty.

 * Fixed changing of expiration date for non-C-99 systems.

 * Fixed a crash while encrypting several files.

 * Fixed a bug while encrypting to several keys.


Noteworthy changes in version 0.7.4 (2006-07-25)
------------------------------------------------

 * Added icon to the Windows version.

 * Other minor fixes.


Noteworthy changes in version 0.7.3 (2006-03-21)
------------------------------------------------

 * Minor fixes.


Noteworthy changes in version 0.7.2 (2006-03-03)
------------------------------------------------

 * The key generation wizard does not allow to set a comment anymore.
   This is an advanced feature available in the advanced GUI version
   of key generation.

 * Bug fixes for the Windows target, in particular
   internationalization and binary mode file handling.


Noteworthy changes in version 0.7.1 (2006-01-09)
------------------------------------------------

 * When verifying the signature on a file, GPA now tries to find
   detached signatures and asks the user whether to verify them.

 * A "refresh" command was added to the keyring. So, if the keyring is
   modified outside GPA (i.e. by reading emails with auto-key-retrieve
   on), you can see the new keys without restarting GPA.

 * A .desktop file for integration with the Gnome and KDE menus is now
   distributed with the tarball.

 * The GPA icon has been changed. The new icon is now used by all
   windows when minimized (and on the window title if supported by the
   window manager).

 * It is again possible to do a build for Windows using the latest
   glib version along with a glib patch as available in the gpg4win
   package.


Noteworthy changes in version 0.7.0 (2003-10-22)
------------------------------------------------

 * Long file operations no longer block GPA, so several operations can be
 run at the same time. This also means GPA does not freeze while an operation
 runs, leading to a more responsive interface.

 * The keyring editor now displays all the subkeys of the currently selected
 key. This is only visible if GPA is in advanced mode (available from the
 preferences dialog).

 * The capabilities of a key (certify, sign, encrypt) are now visible from
 the keyring editor.

 * The keyring editor can now sort keys by any column. By default, they are

src/gpa.c
/* gpa.c - The GNU Privacy Assistant main file.
   Copyright (C) 2000-2002 G-N-U GmbH.
   Copyright (C) 2005, 2008, 2012, 2014 g10 Code GmbH.

   This file is part of GPA

   GPA is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   GPA is distributed in the hope that it will be useful, but WITHOUT
   ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
   or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
   License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, see <http://www.gnu.org/licenses/>.  */

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <signal.h>

#include <glib/gstdio.h>
#include <gtk/gtk.h>

#include <gpgme.h>

#include "get-path.h"
#include "gpa.h"
#include "keymanager.h"
#include "fileman.h"
#include "clipboard.h"
#include "cardman.h"
#include "keyserver.h"
#include "settingsdlg.h"
#include "confdialog.h"
#include "icons.h"

#ifdef __MINGW32__
#include "hidewnd.h"
#endif


/* Global variables. */

/* The home directory of GnuPG.  */
gchar *gnupg_homedir;

/* True if CMS hack mode is enabled.  */
gboolean cms_hack;

/* True if the ticker used for card operations should not be started.  */
gboolean disable_ticker;

/* True if the gpgme edit FSM shall output debug messages.  */
gboolean debug_edit_fsm;

/* True if verbose messages are requested.  */
gboolean verbose;

/* Local variables.  */
typedef struct
{
  gboolean start_key_manager;
  gboolean start_file_manager;
  gboolean start_card_manager;
  gboolean start_clipboard;
  gboolean start_settings;
  gboolean start_only_server;
  gboolean stop_running_server;
  gboolean disable_x509;
  gboolean no_remote;
  gboolean enable_logging;
  gchar *options_filename;
} gpa_args_t;

static char *dummy_arg;

static gpa_args_t args;


/* The copyright notice.  */
static const char *copyright =
"Copyright (C) 2000-2002 Miguel Coca, G-N-U GmbH, Intevation GmbH.\n"
"Copyright (C) 2005-2014 g10 Code GmbH.\n"
"This program comes with ABSOLUTELY NO WARRANTY.\n"
"This is free software, and you are welcome to redistribute it\n"
"under certain conditions.  See the file COPYING for details.\n";


static GtkWidget *backend_config_dialog = NULL;


static void print_version (void);


/* All command line options of the main application.  */
static GOptionEntry option_entries[] =
  {
    { "version", 'V', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK,
      (gpointer) &print_version,
      N_("Output version information and exit"), NULL, },
    { "keyring", 'k', 0, G_OPTION_ARG_NONE, &args.start_key_manager,
      N_("Open key manager (default)"), NULL },
    { "files", 'f', 0, G_OPTION_ARG_NONE, &args.start_file_manager,
      N_("Open file manager"), NULL },
#ifdef ENABLE_CARD_MANAGER
    { "card", 'C', 0, G_OPTION_ARG_NONE, &args.start_card_manager,
      N_("Open the card manager"), NULL },
#endif
    { "clipboard", 'c', 0, G_OPTION_ARG_NONE, &args.start_clipboard,
      N_("Open clipboard"), NULL },
    { "settings", 's', 0, G_OPTION_ARG_NONE, &args.start_settings,
      N_("Open the settings dialog"), NULL },
    { "daemon", 'd', 0, G_OPTION_ARG_NONE, &args.start_only_server,
      N_("Only start the UI server"), NULL },
    { "disable-x509", 0, 0, G_OPTION_ARG_NONE, &args.disable_x509,
      N_("Disable support for X.509"), NULL },
    { "options", 'o', 0, G_OPTION_ARG_FILENAME, &args.options_filename,
      N_("Read options from file"), "FILE" },
    { "no-remote", 0, 0, G_OPTION_ARG_NONE, &args.no_remote,
      N_("Do not connect to a running instance"), NULL },
    { "stop-server", 0, G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE,
      &args.stop_running_server, NULL, NULL },
    /* Note:  the cms option will eventually be removed.  */
    { "cms", 'x', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE,
      &cms_hack, NULL, NULL },
    { "verbose", 'v',      G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE,
      &verbose,  NULL, NULL },
    { "disable-ticker", 0, G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE,
      &disable_ticker, NULL, NULL },
    { "debug-edit-fsm", 0, G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE,
      &debug_edit_fsm, NULL, NULL },
    { "enable-logging", 0, G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE,
      &args.enable_logging, NULL, NULL },
    { "gpg-binary", 0, G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_FILENAME,
      &dummy_arg, NULL, NULL },
    { "gpgsm-binary", 0, G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_FILENAME,
      &dummy_arg, NULL, NULL },
    { NULL }
  };





/* Return a malloced string with the locale dir.  Return NULL on
   error. */
static char *
get_locale_dir (void)
{
#ifdef G_OS_WIN32
  char name[530];
  char *p;

  if (! GetModuleFileNameA (0, name, sizeof (name) - 30))
    return NULL;
  p = strrchr (name, '\\');
  if (!p)
    {
      *name = '\\';
      p = name;
    }
  p++;
  strcpy (p, "share\\locale");
  return strdup (name);
#else
  return strdup (LOCALEDIR);
#endif
}


static void
i18n_init (void)
{
#ifdef ENABLE_NLS
  char *tmp;

  gtk_set_locale ();
  bind_textdomain_codeset (PACKAGE, "UTF-8");

  tmp = get_locale_dir ();
  if (tmp)
    {
      bindtextdomain (PACKAGE, tmp);
      free (tmp);
    }
  textdomain (PACKAGE);
#endif
}


/* Manage the main windows and the settings dialog.  */
static void
quit_if_no_window (void)
{
  if (!args.start_only_server
      && !gpa_key_manager_is_open ()
      && !gpa_file_manager_is_open ()
      && !gpa_clipboard_is_open ()
#ifdef ENABLE_CARD_MANAGER
      && !gpa_card_manager_is_open ()
#endif /*ENABLE_CARD_MANAGER*/
      )
    gpa_stop_server ();
}


static void
close_main_window (GtkWidget *widget, gpointer param)
{
  GtkWidget ** window = param;

  *window = NULL;
  quit_if_no_window ();
}


/* Show the key manager dialog.  */
void
gpa_open_key_manager (GtkAction *action, void *data)
{
  GtkWidget *widget;
  gboolean created;

  widget = gpa_key_manager_get_instance (&created);
  if (created)
    g_signal_connect (G_OBJECT (widget), "destroy",
                      G_CALLBACK (quit_if_no_window), NULL);

  gtk_widget_show_all (widget);
  gtk_window_present (GTK_WINDOW (widget));
}


/* Show the clipboard dialog.  */
void
gpa_open_clipboard (GtkAction *action, void *data)
{
  /* FIXME: Shouldn't this connect only happen if the instance is
     created the first time?  Looks like a memory leak to me.  Right:
     although the closure is ref counted an internal data object will
     get allocated.  */
  g_signal_connect (G_OBJECT (gpa_clipboard_get_instance ()), "destroy",
		    G_CALLBACK (quit_if_no_window), NULL);
  gtk_widget_show_all (gpa_clipboard_get_instance ());

  gtk_window_present (GTK_WINDOW (gpa_clipboard_get_instance ()));
}


/* Show the filemanager dialog.  */
void
gpa_open_filemanager (GtkAction *action, void *data)
{
  /* FIXME: Shouldn't this connect only happen if the instance is
     created the first time?  Looks like a memory leak to me.  */
  g_signal_connect (G_OBJECT (gpa_file_manager_get_instance ()), "destroy",
		    G_CALLBACK (quit_if_no_window), NULL);
  gtk_widget_show_all (gpa_file_manager_get_instance ());

  gtk_window_present (GTK_WINDOW (gpa_file_manager_get_instance ()));
}

/* Show the card manager.  */
#ifdef ENABLE_CARD_MANAGER
void
gpa_open_cardmanager (GtkAction *action, void *data)
{
  /* FIXME: Shouldn't this connect only happen if the instance is
     created the first time?  Looks like a memory leak to me.  */
  g_signal_connect (G_OBJECT (gpa_card_manager_get_instance ()), "destroy",
		    G_CALLBACK (quit_if_no_window), NULL);
  gtk_widget_show_all (gpa_card_manager_get_instance ());

  gtk_window_present (GTK_WINDOW (gpa_card_manager_get_instance ()));
}
#endif /*ENABLE_CARD_MANAGER*/

/* Show the settings dialog.  */
void
gpa_open_settings_dialog (GtkAction *action, void *data)
{
  settings_dlg_new (data);
}


/* Show the backend configuration dialog.  */
void
gpa_open_backend_config_dialog (GtkAction *action, void *data)
{
  if (!backend_config_dialog)
    {
      backend_config_dialog = gpa_backend_config_dialog_new ();
      g_signal_connect (G_OBJECT (backend_config_dialog), "destroy",
			G_CALLBACK (close_main_window), &backend_config_dialog);
      gtk_widget_show_all (backend_config_dialog);
    }

  gtk_window_present (GTK_WINDOW (backend_config_dialog));
}


/* Command line options.  */

/* Print version information and exit.  */
static void
print_version (void)
{
  printf ("%s %s\n%s", PACKAGE, VERSION, copyright);
  exit (0);
}



static void
dummy_log_func (const gchar *log_domain, GLogLevelFlags log_level,
                const gchar *message, gpointer user_data)
{
  /* Nothing to be done.  */
}



/* Helper for main.  */
static gpg_error_t
open_requested_window (int argc, char **argv, int use_server)
{
  gpg_error_t err = 0;
  int did_server = 0;
  int i;

  /* Don't open the key manager if any files are given on the command
     line.  Ditto for the clipboard.  */
  if (args.start_key_manager && (optind >= argc))
    {
      if (use_server)
        {
          did_server = 1;
          err = gpa_send_to_server ("START_KEYMANAGER");
        }
      else
        gpa_open_key_manager (NULL, NULL);
    }

  if (args.start_clipboard && (optind >= argc))
    {
      if (use_server)
        {
          did_server = 1;
          err = gpa_send_to_server ("START_CLIPBOARD");
        }
      else
        gpa_open_clipboard (NULL, NULL);
    }

  if (args.start_file_manager || (optind < argc))
    {
      /* Do not use the server if there are file args - see below.  */
      if (use_server && !(optind < argc))
        {
          did_server = 1;
          err = gpa_send_to_server ("START_FILEMANAGER");
        }
      else
        gpa_open_filemanager (NULL, NULL);
    }

#ifdef ENABLE_CARD_MANAGER
  if (args.start_card_manager)
    {
      if (use_server)
        {
          did_server = 1;
          err = gpa_send_to_server ("START_CARDMANAGER");
        }
      else
        gpa_open_cardmanager (NULL, NULL);
    }
#endif /*ENABLE_CARD_MANAGER*/

  if (args.start_settings)
    {
      if (use_server)
        {
          did_server = 1;
          err = gpa_send_to_server ("START_CONFDIALOG");
        }
      else
        gpa_open_settings_dialog (NULL, NULL);
    }

  /* If there are any command line arguments that are not options, try
     to open them as files in the filemanager.  However if we are to
     connect to a server this can't be done because the running
     instance may already have files in the file manager and thus we
     better do not add new files.  Instead we start a new instance. */
  if (use_server)
    {
      if (!did_server)
        err = -1; /* Create a new instance.  */
    }
  else
    {
      for (i = optind; i < argc; i++)
        gpa_file_manager_open_file (GPA_FILE_MANAGER
                                    (gpa_file_manager_get_instance ()),
                                    argv[i]);
    }

  return err;
}


int
main (int argc, char *argv[])
{
  GError *err = NULL;
  GOptionContext *context;
  char *configname = NULL;
  char *keyservers_configname = NULL;

  /* Under W32 logging is disabled by default to prevent MS Windows NT
     from opening a console.  */
#ifndef G_OS_WIN32
  args.enable_logging = 1;
#endif

  /* Set locale before option parsing for UTF-8 conversion.  */
  i18n_init ();

  /* Parse command line options.  */
  context = g_option_context_new (N_("[FILE...]"));
#if GLIB_CHECK_VERSION (2, 12, 0)
  g_option_context_set_summary (context, N_("Graphical frontend to GnuPG"));
  g_option_context_set_description (context, N_("Please report bugs to <"
						PACKAGE_BUGREPORT ">."));
  g_option_context_set_translation_domain (context, PACKAGE);
#endif
  g_option_context_add_main_entries (context, option_entries, PACKAGE);
  g_option_context_add_group (context, gtk_get_option_group (TRUE));

  if (! g_option_context_parse (context, &argc, &argv, &err))
    {
      g_print ("option parsing failed: %s\n", err->message);
      exit (1);
    }

  if (!args.enable_logging)
    {
#ifdef __MINGW32__
      hide_gpa_console_window();
#endif
      g_log_set_handler ("Glib", G_LOG_LEVEL_CRITICAL
                         | G_LOG_LEVEL_WARNING
                         | G_LOG_LEVEL_MESSAGE
                         | G_LOG_LEVEL_INFO, dummy_log_func, NULL);

      g_log_set_handler ("Gdk", G_LOG_LEVEL_CRITICAL
                         | G_LOG_LEVEL_WARNING
                         | G_LOG_LEVEL_MESSAGE
                         | G_LOG_LEVEL_INFO, dummy_log_func, NULL);

      g_log_set_handler ("Gtk", G_LOG_LEVEL_CRITICAL
                         | G_LOG_LEVEL_WARNING
                         | G_LOG_LEVEL_MESSAGE
                         | G_LOG_LEVEL_INFO, dummy_log_func, NULL);
    }

  gtk_init (&argc, &argv);
#ifdef G_OS_WIN32
  gtk_settings_set_string_property(gtk_settings_get_default(),
                                   "gtk-theme-name",
                                   "MS-Windows",
                                   "XProperty");
#endif

  /* Default icon for all windows.  */
  gtk_window_set_default_icon_from_file (GPA_DATADIR "/gpa.png", &err);
  if (err)
    g_error_free (err);

  gpa_register_stock_items ();

#ifdef IS_DEVELOPMENT_VERSION
  fprintf (stderr, "NOTE: This is a development version!\n");
#endif

  /* Initialize GPGME.  */
  gpgme_check_version (NULL);
#ifdef USE_SIMPLE_GETTEXT
  /* FIXME */
#else
#ifdef ENABLE_NLS
  gpgme_set_locale (NULL, LC_CTYPE, setlocale (LC_CTYPE, NULL));
  gpgme_set_locale (NULL, LC_MESSAGES, setlocale (LC_MESSAGES, NULL));
#endif
#endif

#ifndef G_OS_WIN32
  /* Internationalisation with gtk+-2.0 wants UTF-8 instead of the
     character set determined by locale.  */
  putenv ("OUTPUT_CHARSET=utf8");

  /* We don't want the SIGPIPE.  I wonder why gtk_init does not set it
     to ignore.  Nobody wants this signal.  */
  {
    struct sigaction sa;

    sa.sa_handler = SIG_IGN;
    sigemptyset (&sa.sa_mask);
    sa.sa_flags = 0;
    sigaction (SIGPIPE, &sa, NULL);
  }
#endif

  /* Start the agent if needed.  We need to do this because the card
     manager uses direct assuan commands to the agent and thus expects
     that the agent has been startet. */
  gpa_start_agent ();

  gnupg_homedir = default_homedir ();

  /* GnuPG can not create a key if its home directory is missing.  We
     help it out here.  Should be fixed in GnuPG.  */
  if (! g_file_test (gnupg_homedir, G_FILE_TEST_IS_DIR))
    g_mkdir (gnupg_homedir, 0700);

  /* Locate GPA's configuration file.  */
  if (! args.options_filename)
    configname = g_build_filename (gnupg_homedir, "gpa.conf", NULL);
  else
    configname = args.options_filename;
  gpa_options_set_file (gpa_options_get_instance (), configname);
  g_free (configname);

  if (args.stop_running_server)
    {
      /* If a UI server with the same name is already running, stop
         it.  If not, do nothing.  */
      if (gpa_check_server () == 2)
        gpa_send_to_server ("KILL_UISERVER");
      return 0;
    }

  /* Handle command line options.  */
  cms_hack = !args.disable_x509;

  /* Start the default component.  */
  if (!args.start_key_manager
      && !args.start_file_manager
      && !args.start_clipboard
      && !args.start_settings
      && !args.start_card_manager
      )
    {
      /* The default action is to start the clipboard.  However, if we
         have not yet created a key we remind the user by starting
         with the key manager dialog.  */
      if (key_manager_maybe_firsttime ())
        args.start_key_manager = TRUE;
      else
        args.start_clipboard = TRUE;
    }


  /* Check whether we need to start a server or to simply open a
     window in an already running server.  */
  switch (gpa_check_server ())
    {
    case 0: /* No running server on the expected socket.  Start one.  */
      gpa_start_server ();
      break;
    case 1: /* An old instance or a differen UI server is already running.
               Do not start a server.  */
      break;
    case 2: /* An instance is already running - open the appropriate
               window and terminate.  */
      if (args.no_remote)
        break;
      if (!open_requested_window (argc, argv, 1))
        return 0; /* ready */
      /* Start a new instance on error.  */
      break;
    }

  /* Locate the list of keyservers.  */
  keyservers_configname = g_build_filename (gnupg_homedir, "keyservers", NULL);

  /* Read the list of available keyservers.  */
  keyserver_read_list (keyservers_configname);

  gpa_options_update_default_key (gpa_options_get_instance ());
  /* Now, make sure there are reasonable defaults for the default key
    and keyserver.  */
  if (!is_gpg_version_at_least ("2.1.0")
      && !gpa_options_get_default_keyserver (gpa_options_get_instance ()))
    {
      GList *keyservers = keyserver_get_as_glist ();
      gpa_options_set_default_keyserver (gpa_options_get_instance (),
					 keyservers->data);
    }

  /* Initialize the file watch facility.  */
  gpa_init_filewatch ();

  /* Startup whatever has been requested by the user.  */
  if (!args.start_only_server)
    open_requested_window (argc, argv, 0);

  gtk_main ();

  return 0;
}
